### ISSUE TITLE
Rule proposal: no-external-export-stars to prevent bad scenarios

### ISSUE THREAD

**Issue by dzearing** [typescript-eslint/typescript-eslint#3992]
### Description

When a suite-library has an `index.js` entry which exports 1 or more sub-libraries through `export *`, things become problematic:

1. Duplicate named exports can be easily introduced as sub-library versions change but the suite-library doesn't. When ESM is parsed in the browser where a library may export the same name twice due to export stars, importing that name throws an error.
2. Semver rules are easy to violate. The suite library api surface shifts without a version change, depending on the sub library version.

Libraries should be explicit about what they re-export. This protects the contract from introducing duplicates or shifting without a version change. It's sometimes very non-obvious to developers these situations can happen; especially from packages managed by a single mono-repo by a single owner, but owners change and bad things can happen unintentionally.

We've been considering an eslint rule to enforce this, and as a fixer expand the `export *` to the list of named exports. This ensures there are no named dupes are accidentally exported, and api surface does not change without updating the package version.

If it makes sense to others and seems like a general good practice, I will submit a PR.

### Fail

```ts
export * from 'library';
```

### Pass

```ts
export { Thing1, Thing2, Thing3 } from 'library';
```
